keps/sig-auth/3221-structured-authorization-configuration/README.md

```diff
@@ -0,0 +1,831 @@
+      # This is explicitly used in monitoring machinery for metrics
+      # Note:
+      #   - Validation for this field is similar to how K8s labels are validated today.
+      # Required, with no default
+      name: super-important-kube-system-authorizer
+      # The duration to cache 'authorized' responses from the webhook
+      # authorizer.
+      # Same as setting `--authorization-webhook-cache-authorized-ttl` flag
+      # Default: 5m0s
+      authorizedTTL: 30s
+      # The duration to cache 'authorized' responses from the webhook
+      # authorizer.
+      # Same as setting `--authorization-webhook-cache-unauthorized-ttl` flag
+      # Default: 30s
+      unauthorizedTTL: 30s
+      # Timeout for the webhook request
+      # Maximum allowed is 30s.
+      # Required, with no default.
+      timeout: 3s
+      # The API version of the authorization.k8s.io SubjectAccessReview to
+      # send to and expect from the webhook.
+      # Same as setting `--authorization-webhook-version` flag
+      # Required, with no default
+      # Valid values: v1beta1, v1
+      subjectAccessReviewVersion: v1
+      # Controls the authorization decision when a webhook request fails to
+      # complete or returns a malformed response or errors evaluating
+      # matchConditions.
+      # Valid values:
+      #   - NoOpinion: continue to subsequent authorizers to see if one of
+      #     them allows the request
+      #   - Deny: reject the request without consulting subsequent authorizers
+      # Required, with no default.
+      failurePolicy: Deny
+      connectionInfo:
+        # Controls how the webhook should communicate with the server.
+        # Valid values:
+        # - KubeConfig: use the file specified in kubeConfigFile to locate the
+        #   server.
+        # - InClusterConfig: use the in-cluster configuration to call the
+        #   SubjectAccessReview API hosted by kube-apiserver. This mode is not
+        #   allowed for kube-apiserver.
+        type: KubeConfig
+        # Path to KubeConfigFile for connection info
+        # Required, if connectionInfo.Type is KubeConfig
+        kubeConfigFile: /kube-system-authz-webhook.yaml
+        # matchConditions is a list of conditions that must be met for a request to be sent to this
+        # webhook. An empty list of matchConditions matches all requests.
+        # There are a maximum of 64 match conditions allowed.
+        #
+        # The exact matching logic is (in order):
+        #   1. If at least one matchCondition evaluates to FALSE, then the webhook is skipped.
+        #   2. If ALL matchConditions evaluate to TRUE, then the webhook is called.
+        #   3. If at least one matchCondition evaluates to an error (but none are FALSE):
+        #      - If failurePolicy=Deny, then the webhook rejects the request
+        #      - If failurePolicy=NoOpinion, then the error is ignored and the webhook is skipped
+      matchConditions:
```

Comment:
Sorry for the drive-by review, but a reasonable simplification might be to use CEL `&&` rather than multiple expressions since the semantics described here mirror those provided by CEL. @jpbetz

Reply:
IMO the aesthetics of multiple expressions is much easier to read than `&&`ing a bunch of checks, especially if the checks are complex.

Reply:
@enj the behavior, clarity, and safety should come before aesthetics. In the case, the cost is greater than the benefit. This proposal introduces a second implicit language into the authorization layer on top of CEL which complicates what users must reason about. At first glance, I assumed these expressions were OR'd since that's how many Cloud authorization policies behave. 


The following is strictly less clear than `a && b`

```
expression: a
expression: b
```

And maybe less desirable when considering what happens with the following:

```
expression: a || b && c
expression: d
```

Someone could accidentally assume that this means the same thing as:

```
expression: a || b
expression: c
expression: d
```

However, the correct expansion would be:

```
expression: a || b
expression: a || c
expression: d
```

In short, you'd added much more complexity and risk of an incorrect refactor without much value.

Reply:
> At first glance, I assumed these expressions were OR'd since that's how many Cloud authorization policies behave.

We already use lists of CEL expressions in other Kube APIs, and they are AND'd together.  So for us, this is the most consistent approach.

Reply:
Also the simple examples fail to show the clarity we are going for.  Multiple expressions are far easier to read:

```yaml
- expression: has(request.resourceAttributes)
- expression: request.resourceAttributes.namespace == 'my-ns'
- expression: request.resourceAttributes.name == 'my-name'
```

vs.

```yaml
expression: has(request.resourceAttributes) && request.resourceAttributes.namespace == 'my-ns' && request.resourceAttributes.name == 'my-name'
```

We are also not stopping anyone from using a single large expression if that is more clear to them.

Some other examples of lists of CEL expressions in Kube APIs:

https://github.com/kubernetes/enhancements/tree/master/keps/sig-auth/3331-structured-config-for-oidc-authentication#configuration-file

https://github.com/kubernetes/enhancements/tree/master/keps/sig-api-machinery/3488-cel-admission-control#api-shape